File: PAT_Basic_Level_by_lua/1007 素数对猜想/1007 素数对猜想.lua
-- 题目编号: 1007 素数对猜想
-- 实现原理: 定义素数判断函数（仅检查奇数因子），遍历3到n-2之间的奇数，统计满足x和x+2均为素数的素数对个数。

-- 读取输入的正整数n
local n = tonumber(io.read())

-- 判断一个数是否为素数的函数
local function is_prime(x)
    -- 小于2的数不是素数
    if x < 2 then
        return false
    end
    -- 2是唯一的偶素数
    if x == 2 then
        return true
    end
    -- 偶数（除了2）都不是素数
    if x % 2 == 0 then
        return false
    end
    -- 从3开始，只检查奇数因子，步长为2
    local divisor = 3
    -- 只需检查到sqrt(x)，因为如果x有大于sqrt(x)的因子，则必然有一个对应的小于sqrt(x)的因子
    while divisor * divisor <= x do
        -- 如果能被整除，则不是素数
        if x % divisor == 0 then
            return false
        end
        -- 检查下一个奇数
        divisor = divisor + 2
    end
    -- 没有找到因子，是素数
    return true
end

-- 计数器：统计满足条件的素数对个数
local count = 0
-- 遍历3到n-2之间的所有奇数（步长为2）
-- 只检查奇数，因为除了2以外所有素数都是奇数
for x = 3, n - 2, 2 do
    -- 如果x和x+2都是素数，则构成一个素数对
    if is_prime(x) and is_prime(x + 2) then
        count = count + 1
    end
end
-- 输出素数对的个数
print(count)

File: PAT_Basic_Level_by_c/1007 素数对猜想/1007 素数对猜想.c
/*
 * 题目：1007 素数对猜想
 * 实现原理：
 *   统计不超过n的素数中，差值为2的素数对（即"孪生素数"）的个数。
 *   算法步骤：
 *   1. 实现is_prime函数判断一个数是否为素数：
 *      - 小于等于1不是素数
 *      - 2是唯一的偶素数
 *      - 偶数不是素数
 *      - 从3开始到sqrt(n)，步长为2检查是否能整除
 *   2. 遍历2到n的所有数，记录相邻素数的差值，统计差值为2的次数
 */
#include <stdio.h>
#include <math.h>

int is_prime(int n) {
    if (n <= 1) return 0;                   // 0、1 不是素数
    if (n == 2) return 1;                   // 2 是唯一的偶素数
    if (n % 2 == 0) return 0;               // 其他偶数都不是素数
    for (int i = 3; i <= sqrt(n); i += 2) { // 只检查奇数因子，到 sqrt(n) 为止
        if (n % i == 0) return 0;           // 能被整除说明不是素数
    }
    return 1;
}

int main() {
    int n;
    scanf("%d", &n);
    int count = 0, prev = 2;                // count 统计素数对个数，prev 记录上一个素数
    for (int i = 3; i <= n; i++) {
        if (is_prime(i)) {
            if (i - prev == 2) {            // 与上一个素数相差 2，构成一个素数对
                count++;
            }
            prev = i;                       // 更新上一个素数
        }
    }
    printf("%d\n", count);                  // 输出素数对个数
    return 0;
}